src/sagas/bonding-curve.js

```diff
@@ -0,0 +1,363 @@
+  )
+
+  return { totalETH, totalPNK, allowance }
+}
+
+/**
+ * Buy PNK from the bonding curve.
+ * @param {string} amount The amount of ETH the user has input.
+ */
+function* buyPNKFromBondingCurve({ payload: { amount } }) {
+  yield put(bondingCurveActions.setUpdating(true))
+  const addr = yield select(walletSelectors.getAccount)
+  const success = yield call(
+    getBondingCurve().buy,
+    addr,
+    1,
+    '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
+    amount,
+    addr
+  )
+  yield put(bondingCurveActions.setUpdating(false))
+
+  if (!success) return
+  yield call(getBondingCurve().waitForBuy, addr)
+  yield all([
+    put(arbitratorActions.fetchPNKBalance()),
+    put(bondingCurveActions.fetchBondingCurveData()),
+    put(walletActions.fetchBalance())
+  ])
+}
+
+/**
+ * Sell PNK to the bonding curve.
+ * @param {string} amount The amount of PNK the user has input.
+ */
+function* sellPNKToBondingCurve({ payload: { amount } }) {
+  yield put(bondingCurveActions.setUpdating(true))
+  const addr = yield select(walletSelectors.getAccount)
+  const success = yield call(
+    getBondingCurve().sell,
+    amount,
+    addr,
+    1,
+    '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
+    addr
+  )
+  yield put(bondingCurveActions.setUpdating(false))
+  if (!success) return
+  yield call(getBondingCurve().waitForSell, addr)
+  yield all([
+    put(arbitratorActions.fetchPNKBalance()),
+    put(bondingCurveActions.fetchBondingCurveData()),
+    put(walletActions.fetchBalance())
+  ])
+}
+
+/**
+ * Approve PNK to the bonding curve contract.
+ */
+function* approvePNKToBondingCurve({ payload: { amount } }) {
+  yield put(bondingCurveActions.updateApproveTransactionProgress('pending'))
+
+  const addr = yield select(walletSelectors.getAccount)
+  const txID = yield call(getBondingCurve().approve, amount, addr)
+
+  if (!txID) {
+    yield put(bondingCurveActions.updateApproveTransactionProgress(''))
+    return
+  }
+
+  yield put(bondingCurveActions.updateApproveTransactionProgress('confirming'))
+  yield call(getBondingCurve().waitForApproval, addr)
+
+  yield put(bondingCurveActions.updateApproveTransactionProgress('done'))
+  yield call(fetchBondingCurveTotals)
+}
+
+/**
+ * The root of the bonding curve saga.
+ */
+export default function* bondingCurveSaga() {
+  yield takeLatest(
+    bondingCurveActions.bondingCurve.FETCH,
+    lessduxSaga,
+    'fetch',
+    bondingCurveActions.bondingCurve,
+    fetchBondingCurveTotals
+  )
+  yield takeLatest(
+    bondingCurveActions.bondingCurve.BUY_PNK,
+    buyPNKFromBondingCurve
+  )
+  yield takeLatest(
+    bondingCurveActions.bondingCurve.SELL_PNK,
+    sellPNKToBondingCurve
+  )
+  yield takeLatest(
+    bondingCurveActions.bondingCurve.APPROVE_PNK,
+    approvePNKToBondingCurve
+  )
+}
+
+/**
+ * Provides interaction with the bonding curve contract on the blockchain.
+ */
+class BondingCurve extends ContractImplementation {
```

Comment:
This seems out of place here. Move to separate file?

Reply:
Also feels a little unnecessary to create a whole ContractImplementation for BondingCurve. Why not just call the functions directly from the web3 contract instance in the generators? You have the JSON interface in the app, and in pretty much all of your functions you aren't doing anything besides just calling the function on the contractInstance. Seems like a lot of extra code just to pass through the call

Reply:
This is indeed a very thin layer but IMO having it helps keeping the generators uncluttered. Plus this serves as a document of the parameters passed down to the contract.